core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/iterator/JoinIterator.java

```diff
@@ -61,7 +68,15 @@ protected BindingSet getNextElement() throws QueryEvaluationException {
 				rightIter.close();
 
 				if (leftIter.hasNext()) {
-					rightIter = strategy.evaluate(join.getRightArg(), leftIter.next());
+					TupleExpr rightArg = join.getRightArg();
+					if (TupleExprs.isGraphPatternGroup(rightArg) && !TupleExprs.isFilterExistsFunction(rightArg)) {
+						// leftiter bindings are out of scope for the right arg, so we merge afterward.
+						BindingSet next = leftIter.next();
+						rightIter = new MergeIteration(next, new BindingSetFilterIteration(next,
```

Comment:
I did a bit of debugging for the no longer working query pattern from org.eclipse.rdf4j.sail.federation.FederationQueryTest.test(), i.e.

```
{ ?person b:name ?name ; b:desc ?desc }
```

What I could observe by adding the following additional debugging code in line 73:

```
System.out.println(Iterations.asList(
	strategy.evaluate(rightArg, new EmptyBindingSet())));
```

and moreover the following debugging output between after line 74

```
System.out.println(Iterations.asList(
	new BindingSetFilterIteration(next,
	strategy.evaluate(rightArg, new EmptyBindingSet()))));
```

I see that we are loosing the second relevant bindingset from the inner join because of the BindingSetFilterIteration.

I am not sure if it is correct at all in the first place that this Join is interpreted as GraphPatternGroup, my feeling is that this is already the problem. However, I do not yet understand what GraphPatternGroups really represent.

Does this help to get a step further?

Reply:
It definitely does, thanks. And you're probably right about the cause as well. A graph pattern group is mostly a syntactical construct so it has no real place in the algebra. However, the SPARQL algebra is based on bottom-up evaluation while RDF4J's engine typically uses a more top-down approach. This is mostly not a problem but in some cases (such as this one) it messes with variable scoping, which is why we have some exception cases built-in. 

I'll take a closer look at why this is marked as a group pattern and see if we can somehow bypass that without the code becoming too obscure.

Reply:
I've discovered something relevant: in breaking down the query to send to various federation members, the Federation sail produces this:
```
SELECT ?name ?person ?desc
WHERE {
{?person <urn:schema:b:name> ?name  .}
{?person <urn:schema:b:desc> ?desc  .}
}
```

Notice the extra braces around each statement pattern. It's this that causes the parser to mark each pattern as a separate group graph pattern. I'll dig a bit deeper to find out why it's necessary to produce the query in this way. Perhaps these extra braces can be eliminated.